Jordyn Fuchs

# Homework 3

### Problem 3

Train a binary logistic regression model using mini-batch SGD. Use the logistic regression model.

In [1]:
# import packages

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [16]:
# create sigmoid function

def sigmoid(z):
    return 1 / (1 + np.exp(-z))

# create cross-entropy loss function

def crossentropy(t, y):
    epsilon = 1e-5
    return (-1 / len(t)) * (np.sum( t * np.log(y + epsilon) + (1 - t) * np.log( 1 - y + epsilon ) ))

# create logistic regression function

def logisticregression(X, t, batchsize = 64, learningrate = 0.01, iterations = 100):
    row, col = X.shape
    w = np.zeros(col)

    for i in range(0, row, batchsize):
        X_sub = X[i:i+batchsize]
        t_sub = t[i:i+batchsize]

        y = sigmoid(np.dot(X_sub, w))
        gradient = np.dot((y - t_sub), X_sub) * (1/batchsize)

        w = w - (gradient * learningrate)

    return w


### Problem 4a

Download Wisconsin Breast Cancer dataset.

In [5]:
# read in dataset

#!pip install ucimlrepo

from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
breast_cancer_wisconsin_diagnostic = fetch_ucirepo(id=17) 
  
# data (as pandas dataframes) 
X = breast_cancer_wisconsin_diagnostic.data.features 
y = breast_cancer_wisconsin_diagnostic.data.targets 
  
# metadata 
print(breast_cancer_wisconsin_diagnostic.metadata) 
  
# variable information 
print(breast_cancer_wisconsin_diagnostic.variables) 


{'uci_id': 17, 'name': 'Breast Cancer Wisconsin (Diagnostic)', 'repository_url': 'https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic', 'data_url': 'https://archive.ics.uci.edu/static/public/17/data.csv', 'abstract': 'Diagnostic Wisconsin Breast Cancer Database.', 'area': 'Health and Medicine', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 569, 'num_features': 30, 'feature_types': ['Real'], 'demographics': [], 'target_col': ['Diagnosis'], 'index_col': ['ID'], 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 1993, 'last_updated': 'Fri Nov 03 2023', 'dataset_doi': '10.24432/C5DW2B', 'creators': ['William Wolberg', 'Olvi Mangasarian', 'Nick Street', 'W. Street'], 'intro_paper': {'ID': 230, 'type': 'NATIVE', 'title': 'Nuclear feature extraction for breast tumor diagnosis', 'authors': 'W. Street, W. Wolberg, O. Mangasarian', 'venue': 'Electronic imaging', 'year': 1993, 'journal': None, 'DOI': '1

### Problem 4b
Split data into train, validation, and test sets.

In [6]:
# use scikit-learn to split data into train, validation, and test sets
# 80% train, 10% validation, 10% test

X_train, X_test_full, y_train, y_test_full = train_test_split(X, y, 
test_size=0.2, random_state = 28)
X_val, X_test, y_val, y_test = train_test_split(X_test_full, y_test_full, 
test_size = 0.5, random_state = 28)

### Problem 4c
Report the size of each class in your training and validation set.

In [10]:
print(y_train.value_counts())
print(y_val.value_counts())

Diagnosis
B            284
M            171
Name: count, dtype: int64
Diagnosis
B            40
M            17
Name: count, dtype: int64


In the training set, there are 284 benign values and 171 malignant values. In the validation set, there are 40 benign values and 17 malignant values.

### Problem 4d
Train a binary logistic regression model using your implementation from problem 3. Initialize the model weights randomly, sampling from a standard Gaussian distribution.

In [19]:
# use logisticregression function created in problem 3

w = logisticregression(X_train, y_train, batchsize = 16, learningrate = 0.1, iterations = 100)
print(w)

w2 = logisticregression(X_train, y_train, batchsize = 32, learningrate = 0.01, iterations = 100)
print(w2)

w3 = logisticregression(X_train, y_train, batchsize = 64, learningrate = 0.001, iterations = 100)
print(w3)

Unexpected exception formatting exception. Falling back to standard exception


Traceback (most recent call last):
  File "/Users/jordynfuchs/anaconda3/envs/pic16b/lib/python3.11/site-packages/IPython/core/interactiveshell.py", line 3553, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "/var/folders/lg/rf0x35w16dzdlc87h7c337g40000gn/T/ipykernel_16890/4193081640.py", line 3, in <module>
    w = logisticregression(X_train, y_train, batchsize = 16, learningrate = 0.1, iterations = 100)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/folders/lg/rf0x35w16dzdlc87h7c337g40000gn/T/ipykernel_16890/2902290470.py", line 23, in logisticregression
    gradient = np.dot((y - t_sub), X_sub) * (1/batchsize)
                       ~~^~~~~~~
  File "/Users/jordynfuchs/anaconda3/envs/pic16b/lib/python3.11/site-packages/pandas/core/generic.py", line 2102, in __array_ufunc__
    return arraylike.array_ufunc(self, ufunc, method, *inputs, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In [23]:
# instantiate the model (using the default parameters)
logreg = LogisticRegression(random_state=28)

# fit the model with data
logreg.fit(X_train, y_train)

ypred = logreg.predict(X_test)

/Users/jordynfuchs/anaconda3/envs/pic16b/lib/python3.11/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/jordynfuchs/anaconda3/envs/pic16b/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### Problem 4e
Use the trained model to report the performance on the model test set. Use accuracy, precision, recall, and F1-score for evaluation metrics.

In [22]:
# get y predictions using test set

y_prob = sigmoid(np.dot(X_test, w))
y_pred = (y_prob >= 0.5).astype(int)

NameError: name 'w' is not defined

In [33]:
from sklearn import metrics
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

confusion = metrics.confusion_matrix(y_test, ypred)
print("confusion matrix: ", confusion)

# accuracy
accuracy = accuracy_score(y_test, ypred)
print("accuracy: ", accuracy)

# precision
precision = 23/(23+3)
print("precision: ", precision)

# recall
recall = precision = 23/(23+1)
print("recall: ", recall)

# f1 score
print("f1: ", ((2*precision*recall/(precision + recall))))

confusion matrix:  [[30  3]
 [ 1 23]]
accuracy:  0.9298245614035088
precision:  0.8846153846153846
recall:  0.9583333333333334
f1:  0.9583333333333334


### Problem 4f
Summarize your findings.

The model has a relatively high accuracy of 92.98%. There is a higher recall (95.83%) than precision (88.46%), meaning that it is effective at identifying the positive cases but it is also resulting in some false positive values. However, both metrics are still relatively high indicating strong results. Finally, the model has a high f1 score of 95.83%, which indicates a strong balance between precision and recall.